service/history/historyEngine.go

```diff
@@ -1593,6 +1582,207 @@ func (e *historyEngineImpl) SignalWorkflowExecution(signalRequest *h.SignalWorkf
+	*workflow.StartWorkflowExecutionResponse, error) {
+
+	domainID, err := getDomainUUID(signalWithStartRequest.DomainUUID)
+	if err != nil {
+		return nil, err
+	}
+	sRequest := signalWithStartRequest.SignalWithStartRequest
+	execution := workflow.WorkflowExecution{
+		WorkflowId: sRequest.WorkflowId,
+	}
+
+	isBrandNew := false
+	prevRunID := ""
+	attempt := 0
+
+	context, release, err0 := e.historyCache.getOrCreateWorkflowExecution(domainID, execution)
+
+	if err0 == nil {
+		defer release()
+	Just_Signal_Loop:
+		for ; attempt < conditionalRetryCount; attempt++ {
+			msBuilder, err1 := context.loadWorkflowExecution()
+			if err1 != nil {
+				if _, ok := err1.(*workflow.EntityNotExistsError); ok {
+					isBrandNew = true
+					break
+				}
+				return nil, err1
+			}
+			if !msBuilder.isWorkflowExecutionRunning() {
+				prevRunID = context.workflowExecution.GetRunId()
+				break
+			}
+
+			if msBuilder.AddWorkflowExecutionSignaled(getSignalRequest(sRequest)) == nil {
+				return nil, &workflow.InternalServiceError{Message: "Unable to signal workflow execution."}
+			}
+
+			var transferTasks []persistence.Task
+			var timerTasks []persistence.Task
+			// Create a transfer task to schedule a decision task
+			if !msBuilder.HasPendingDecisionTask() {
+				di := msBuilder.AddDecisionTaskScheduledEvent()
+				transferTasks = append(transferTasks, &persistence.DecisionTask{
+					DomainID:   domainID,
+					TaskList:   di.Tasklist,
+					ScheduleID: di.ScheduleID,
+				})
+				if msBuilder.isStickyTaskListEnabled() {
+					tBuilder := e.getTimerBuilder(&context.workflowExecution)
+					stickyTaskTimeoutTimer := tBuilder.AddScheduleToStartDecisionTimoutTask(di.ScheduleID, di.Attempt,
+						msBuilder.executionInfo.StickyScheduleToStartTimeout)
+					timerTasks = append(timerTasks, stickyTaskTimeoutTimer)
+				}
+			}
+			// Generate a transaction ID for appending events to history
+			transactionID, err2 := e.shard.GetNextTransferTaskID()
+			if err2 != nil {
+				return nil, err2
+			}
+
+			// We apply the update to execution using optimistic concurrency.  If it fails due to a conflict then reload
+			// the history and try the operation again.
+			if err := context.updateWorkflowExecution(transferTasks, timerTasks, transactionID); err != nil {
+				if err == ErrConflict {
+					continue Just_Signal_Loop
+				}
+				return nil, err
+			}
+			e.timerProcessor.NotifyNewTimer(timerTasks)
+			return &workflow.StartWorkflowExecutionResponse{RunId: context.workflowExecution.RunId}, nil
+		} // end for Just_Signal_Loop
+		if attempt == conditionalRetryCount {
+			return nil, ErrMaxAttemptsExceeded
+		}
+	} else {
+		if _, ok := err0.(*workflow.EntityNotExistsError); ok {
+			isBrandNew = true
+		} else {
+			return nil, err0
+		}
+	}
+
+	// Start workflow and signal
+	request := getStartRequest(sRequest)
+	if request.ExecutionStartToCloseTimeoutSeconds == nil || *request.ExecutionStartToCloseTimeoutSeconds <= 0 {
```

Comment:
maybe worth to reuse those parameter check with StartWorkflow()

Reply:
good suggestion, changed.